File: Hub/src/main/java/com/salcedo/rapbot/hub/Hub.java
package com.salcedo.rapbot.hub;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.Terminated;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import com.salcedo.rapbot.driver.DriverActor;
import com.salcedo.rapbot.driver.KeyboardDriverStrategy;
import com.salcedo.rapbot.locomotion.MotorActor;
import com.salcedo.rapbot.sense.SenseActor;
import com.salcedo.rapbot.snapshot.RegisterSubSystemMessage;
import com.salcedo.rapbot.snapshot.SnapshotRouterActor;
import com.salcedo.rapbot.snapshot.StartSnapshotMessage;
import com.salcedo.rapbot.snapshot.SystemSnapshot;
import com.salcedo.rapbot.userinterface.GraphicalUserInterface;
import com.salcedo.rapbot.userinterface.GraphicalUserInterfaceActor;
import com.salcedo.rapbot.vision.VisionActor;

import java.nio.file.Path;

public final class Hub extends AbstractActor {
    private final LoggingAdapter log = Logging.getLogger(this);
    private final ServiceFactory serviceFactory;
    private final GraphicalUserInterface gui;
    private final Path workingDirectory;

    public Hub(
            final ServiceFactory serviceFactory,
            final GraphicalUserInterface gui,
            final Path workingDirectory) {
        this.serviceFactory = serviceFactory;
        this.gui = gui;
        this.workingDirectory = workingDirectory;
    }

    public static Props props(
            final ServiceFactory serviceFactory,
            final GraphicalUserInterface gui,
            final Path workingDirectory) {
        return Props.create(Hub.class, serviceFactory, gui, workingDirectory);
    }

    @Override
    public void preStart() {
        log.info("Starting Hub with working directory of: {}.", workingDirectory);

        final ActorRef motors = getContext().actorOf(MotorActor.props(serviceFactory.motor()), "motors");
        final ActorRef vision = getContext().actorOf(VisionActor.props(serviceFactory.vision()), "vision");
        final ActorRef sensors = getContext().actorOf(SenseActor.props(serviceFactory.sense()), "sensors");
        final ActorRef driver = getContext().actorOf(DriverActor.props(motors, new KeyboardDriverStrategy()), "driver");

        //getContext().actorOf(SnapshotWriterActor.props(workingDirectory), "writer");
        getContext().actorOf(GraphicalUserInterfaceActor.props(gui), "gui");

        createSnapshot();

        context().system().eventStream().publish(new RegisterSubSystemMessage(vision));
        context().system().eventStream().publish(new RegisterSubSystemMessage(sensors));
        context().system().eventStream().publish(new RegisterSubSystemMessage(driver));
        context().system().eventStream().publish(new RegisterSubSystemMessage(motors));
        context().system().eventStream().publish(new StartSnapshotMessage());
    }

    private void createSnapshot() {
        final ActorRef snapshot = getContext().actorOf(SnapshotRouterActor.props(), "snapshot");

        context().system().eventStream().subscribe(snapshot, StartSnapshotMessage.class);
        context().system().eventStream().subscribe(snapshot, RegisterSubSystemMessage.class);
        context().system().eventStream().subscribe(self(), SystemSnapshot.class);
    }

    @Override
    public Receive createReceive() {
        return receiveBuilder()
                .match(Terminated.class, this::terminate)
                .match(SystemSnapshot.class, this::state)
                .build();
    }

    private void state(final SystemSnapshot snapshot) {
        getContext().getSystem().eventStream().publish(new SnapshotBackedSystemState(snapshot, context()));
    }

    private void terminate(Terminated message) {
        log.error("Actor terminated: {}. Shutting down system.", message.actor());
    }
}

File: Hub/src/main/scala/com/salcedo/rapbot/hub/Hub.scala
package com.salcedo.rapbot.hub

import java.time.Instant

import akka.actor.Status.Success
import akka.actor.{Actor, ActorLogging, ActorRef, Props, Terminated}
import com.salcedo.rapbot.driver.DriverActor.Drive
import com.salcedo.rapbot.hub.Hub.{SubSystem, SystemState}
import com.salcedo.rapbot.motor.MotorActor.Vehicle
import com.salcedo.rapbot.snapshot.SnapshotActor.Snapshot
import com.salcedo.rapbot.snapshot.SnapshotTakerActor
import com.salcedo.rapbot.snapshot.SnapshotTakerActor.{RegisterSubSystem, TakeSnapshot}
import com.salcedo.rapbot.vision.VisionActor.StillFrame

object Hub {

  def props(subSystems: SubSystem*): Props = {
    Props(new Hub(subSystems))
  }

  case class SubSystem(props: Props, name: String)

  case class SystemState(
                          trigger: String,
                          uuid: String,
                          start: Long,
                          timeWindow: Long,
                          drive: Drive,
                          vehicle: Vehicle,
                          imagePath: String
                        )

}

class Hub(subSystems: Seq[SubSystem]) extends Actor with ActorLogging {
  private val snapshot = context.actorOf(SnapshotTakerActor.props, "snapshot")
  var children: Set[ActorRef] = subSystems.map(actorFor).toSet

  override def preStart(): Unit = {
    context.system.eventStream.subscribe(self, classOf[Snapshot])

    children.foreach(snapshot ! RegisterSubSystem(_))
    snapshot ! TakeSnapshot(self.path.name)
  }

  override def receive: PartialFunction[Any, Unit] = {
    case terminated: Terminated => this.terminate(terminated)
    case snapshot: Snapshot => this.state(snapshot)
  }

  private def state(snapshot: Snapshot): Unit = {
    val snapshots = snapshot.responses.values
      .filter(_.isInstanceOf[Success])
      .map(_.asInstanceOf[Success])
      .map(_.status)

    val drive = snapshots.find(_.isInstanceOf[Drive]).map(_.asInstanceOf[Drive])
    val vehicle: Option[Vehicle] = snapshots.find(_.isInstanceOf[Vehicle]).map(_.asInstanceOf[Vehicle])
    val image: Option[StillFrame] = snapshots.find(_.isInstanceOf[StillFrame]).map(_.asInstanceOf[StillFrame])

    context.system.eventStream.publish(SystemState(
      snapshot.trigger,
      snapshot.uuid.toString,
      snapshot.start.toEpochMilli,
      snapshot.duration.toMillis,
      drive.orNull,
      vehicle.orNull,
      image.map(_.path).getOrElse("/dev/null")
    ))
  }

  private def terminate(message: Terminated): Unit = {
    log.error("Actor terminated: {}.", message.actor)

    children -= message.actor
  }

  def actorFor(system: SubSystem): ActorRef = {
    context.watch(context.actorOf(system.props, system.name))
  }
}

